### PR TITLE
fix: handle tags including numbers in `prepare` command

### LINKED ISSUE #164: `corepack prepare` doesn't work for tags including a number
`corepack prepare npm@latest` works but `corepack prepare npm@latest-7` doesn't.

The result in corepack@0.13.0:

```
$ corepack prepare npm@latest-7
Type Error: Invalid comparator: latest-7
    at Comparator.parse (/Users/teppeis/.nodenv/versions/18.7.0/lib/node_modules/corepack/dist/corepack.js:3370:13)
    at new Comparator (/Users/teppeis/.nodenv/versions/18.7.0/lib/node_modules/corepack/dist/corepack.js:3354:10)
    at /Users/teppeis/.nodenv/versions/18.7.0/lib/node_modules/corepack/dist/corepack.js:3611:47
    at Array.map (<anonymous>)
    at Range.parseRange (/Users/teppeis/.nodenv/versions/18.7.0/lib/node_modules/corepack/dist/corepack.js:3611:35)
    at /Users/teppeis/.nodenv/versions/18.7.0/lib/node_modules/corepack/dist/corepack.js:3512:22
    at Array.map (<anonymous>)
    at new Range (/Users/teppeis/.nodenv/versions/18.7.0/lib/node_modules/corepack/dist/corepack.js:3512:8)
    at Module.findInstalledVersion (/Users/teppeis/.nodenv/versions/18.7.0/lib/node_modules/corepack/dist/corepack.js:15788:19)
    at async Engine.resolveDescriptor (/Users/teppeis/.nodenv/versions/18.7.0/lib/node_modules/corepack/dist/corepack.js:15214:31)
(node:45466) Warning: Closing directory handle on garbage collection
(Use `node --trace-warnings ...` to show where the warning was created)
```

This regex seems to be the cause.
https://github.com/nodejs/corepack/blob/f241dff8d5505faeba5f38354b97006305cb6929/sources/Engine.ts#L139

### PR DESCRIPTION
**Pull request by teppeis**
fix #164

This is based on the detection logic in npm cli.
https://github.com/npm/npm-package-arg/blob/573b4b80d37dc3976d1b3b1a972d95b65e0d27c2/lib/npa.js#L389-L400

### CODE REVIEW

**Review by aduh95 — APPROVED** [review]

**Comment by aduh95** [comment]
Do you think we could add a test for that?

**Comment by teppeis** [comment]
@aduh95 OK, I'll try.

It seems that CI tests are failing with high frequency even in the main branch.
Do you know the cause?

**Comment by teppeis** [comment]
@aduh95 Thank you for https://github.com/nodejs/corepack/pull/168.
After it is merged I'd like to add tests for this PR.
If that is going to take some time, please merge this PR first.

**Comment by aduh95** [comment]
> @aduh95 Thank you for #168. After it is merged I'd like to add tests for this PR. If that is going to take some time, please merge this PR first.

Please don't wait for #168, the faster you add tests, the faster this can be merged, regardless of the flakiness of the CI :)

**Comment by teppeis** [comment]
@aduh95 I've added a test.

File: tests/main.test.ts [hunk 1]
```diff
@@ -208,6 +208,25 @@ it(`should allow updating the pinned version using the "prepare" command`, async
...
+      exitCode: 0,
+      stderr: ``,
+    });
+
+    await xfs.writeJsonPromise(ppath.join(cwd, `package.json` as Filename), {
+      // empty package.json file
+    });
+
+    await expect(runCli(cwd, [`npm`, `--version`])).resolves.toMatchObject({
+      stdout: `7.24.2\n`,
```
Comment (aduh95) [inline tests/main.test.ts:223]: That's going to fail next time npm updates the `latest-7` tag, are we positive it's not going to change? If so, let's add a comment, ideally with a link to their announcement.

Comment (teppeis) [reply tests/main.test.ts:223]: Yes, `latest-7` will be updated.
However, I thought it was OK because the communication was recorded by nock. Is that not sufficient, and should I make sure it succeeds even if nock replay is disabled?

Comment (aduh95) [reply tests/main.test.ts:223]: On my own machine, I don't always put `NOCK_ENV=replay` in my env, so the test would start failing in that case. Hopefully we can come up with a more robust test. What do you think of

```js
    const npmVersion = await runCli(cwd, [`npm`, `--version`]);
    expect(npmVersion).toMatchObject({
      stderr: ``,
      exitCode: 0,
    });
    expect(npmVersion.stdout).toMatch(/^7\./);
```

Comment (teppeis) [reply tests/main.test.ts:223]: That's a reasonable solution!

File: tests/main.test.ts [hunk 2]
```diff
@@ -1,4 +1,5 @@
 import {Filename, ppath, xfs, npath, PortablePath} from '@yarnpkg/fslib';
+import exp                                         from 'constants';
```
Comment (aduh95) [inline tests/main.test.ts:2]: Unrelated change?
```suggestion
```

Comment (teppeis) [reply tests/main.test.ts]: oh, thank you.

**Review by aduh95 — APPROVED** [review]

File: tests/main.test.ts (hunk 1 above)
Comment (teppeis) [reply tests/main.test.ts]: @merceyz Thank you! I'll fix it.

Comment (teppeis) [reply tests/main.test.ts]: https://github.com/nodejs/corepack/pull/176

**Review by Coryhotkiss — COMMENTED** [review]
I want it all set up

**Review by Assylumware — COMMENTED** [review]
I want it all set
